source/base/polynomials_rt_bubbles.cc

```diff
@@ -0,0 +1,524 @@
+         ExcDimensionMismatch(third_derivatives.size(), n_pols));
+  Assert(fourth_derivatives.size()==n_pols || fourth_derivatives.size()==0,
+         ExcDimensionMismatch(fourth_derivatives.size(), n_pols));
+
+  // Third and fourth derivatives are not implemented
+  (void)third_derivatives;
+  Assert(third_derivatives.size()==0,
+         ExcNotImplemented());
+  (void)fourth_derivatives;
+  Assert(fourth_derivatives.size()==0,
+         ExcNotImplemented());
+
+  const unsigned int n_sub = polynomial_space.n();
+
+  // guard access to the scratch arrays in the following block
+  // using a mutex to make sure they are not used by multiple threads
+  // at once
+
+  static Threads::Mutex mutex;
+  Threads::Mutex::ScopedLock lock(mutex);
+
+  static std::vector<double> p_values;
+  static std::vector<Tensor<1,dim> > p_grads;
+  static std::vector<Tensor<2,dim> > p_grad_grads;
+  static std::vector<Tensor<3,dim> > p_third_derivatives;
+  static std::vector<Tensor<4,dim> > p_fourth_derivatives;
+
+  p_values.resize((values.size() == 0) ? 0 : n_sub);
+  p_grads.resize((grads.size() == 0) ? 0 : n_sub);
+  p_grad_grads.resize((grad_grads.size() == 0) ? 0 : n_sub);
+
+  // This is the Raviart-Thomas part of the space
+  for (unsigned int d=0; d<dim; ++d)
+    {
+      // First we copy the point. The
+      // polynomial space for
+      // component d consists of
+      // polynomials of degree k+1 in
+      // x_d and degree k in the
+      // other variables. in order to
+      // simplify this, we use the
+      // same AnisotropicPolynomial
+      // space and simply rotate the
+      // coordinates through all
+      // directions.
+      Point<dim> p;
+      for (unsigned int c=0; c<dim; ++c)
+        p(c) = unit_point((c+d)%dim);
+
+      polynomial_space.compute (p, p_values, p_grads, p_grad_grads, p_third_derivatives, p_fourth_derivatives);
+
+      for (unsigned int i=0; i<p_values.size(); ++i)
+        values[i+d*n_sub][d] = p_values[i];
+
+      for (unsigned int i=0; i<p_grads.size(); ++i)
+        for (unsigned int d1=0; d1<dim; ++d1)
+          grads[i+d*n_sub][d][(d1+d)%dim] = p_grads[i][d1];
+
+      for (unsigned int i=0; i<p_grad_grads.size(); ++i)
+        for (unsigned int d1=0; d1<dim; ++d1)
+          for (unsigned int d2=0; d2<dim; ++d2)
+            grad_grads[i+d*n_sub][d][(d1+d)%dim][(d2+d)%dim]
+              = p_grad_grads[i][d1][d2];
+
+
+    }
+
+  // Next we compute the polynomials and derivatives
+  // of the curl part of the space
+  std::vector<std::vector<double> > monoval_plus(dim, std::vector<double>(4));
+  std::vector<std::vector<double> > monoval(dim, std::vector<double>(4));
+
+  std::vector<std::vector<double> > monoval_i(dim, std::vector<double>(4));
+  std::vector<std::vector<double> > monoval_j(dim, std::vector<double>(4));
+  std::vector<std::vector<double> > monoval_jplus(dim, std::vector<double>(4));
+
+  unsigned int start = dim*n_sub;
+
+  for (unsigned int d=0; d<dim; ++d)
```

Comment:
Can you add a comment with the formula of the functions you are adding. It's hard to check that it is correct without it.